homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,266 @@
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN, KEY_COORDINATOR, KEY_INVERTER
+
+# Service related constants
+SERVICE_SET_WORK_MODE = "set_work_mode"
+ATTR_WORK_MODE = "work_mode"
+SET_WORK_MODE_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_WORK_MODE): cv.positive_int,
+    }
+)
+SERVICE_SET_ONGRID_BATTERY_DOD = "set_ongrid_battery_dod"
+ATTR_ONGRID_BATTERY_DOD = "ongrid_battery_dod"
+SET_ONGRID_BATTERY_DOD_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ONGRID_BATTERY_DOD): cv.positive_int,
+    }
+)
+SERVICE_SET_GRID_EXPORT_LIMIT = "set_grid_export_limit"
+ATTR_GRID_EXPORT_LIMIT = "grid_export_limit"
+SET_GRID_EXPORT_LIMIT_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_GRID_EXPORT_LIMIT): cv.positive_int,
+    }
+)
+
+_MAIN_SENSORS = (
+    "house_consumption",
+    "active_power",
+    "battery_soc",
+    "e_day",
+)
+
+_ICONS = {
+    SensorKind.PV: "mdi:solar-power",
+    SensorKind.AC: "mdi:power-plug-outline",
+    SensorKind.UPS: "mdi:power-plug-off-outline",
+    SensorKind.BAT: "mdi:battery-high",
+    SensorKind.GRID: "mdi:transmission-tower",
+}
+
+_DEVICE_CLASSES = {
+    "A": DEVICE_CLASS_CURRENT,
+    "V": DEVICE_CLASS_VOLTAGE,
+    "W": DEVICE_CLASS_POWER,
+    "kWh": DEVICE_CLASS_ENERGY,
+    "C": DEVICE_CLASS_TEMPERATURE,
+    "Hz": DEVICE_CLASS_VOLTAGE,
+}
+
+_STATE_CLASSES = {
+    "A": STATE_CLASS_MEASUREMENT,
+    "V": STATE_CLASS_MEASUREMENT,
+    "W": STATE_CLASS_MEASUREMENT,
+    "kWh": STATE_CLASS_TOTAL_INCREASING,
+    "C": STATE_CLASS_MEASUREMENT,
+    "Hz": STATE_CLASS_MEASUREMENT,
+}
+
+_UNITS = {
+    "A": ELECTRIC_CURRENT_AMPERE,
+    "V": ELECTRIC_POTENTIAL_VOLT,
+    "W": POWER_WATT,
+    "kWh": ENERGY_KILO_WATT_HOUR,
+    "C": TEMP_CELSIUS,
+    "Hz": FREQUENCY_HERTZ,
+}
+
+
+def _get_sensor_description(sensor):
+    """Create entity description for specified inverter sensor."""
+    desc = SensorEntityDescription(
+        key=sensor.id_,
+        name=sensor.name.strip(),
+        icon=_ICONS.get(sensor.kind),
+        native_unit_of_measurement=_UNITS.get(sensor.unit, sensor.unit),
+        device_class=_DEVICE_CLASSES.get(sensor.unit),
+        state_class=_STATE_CLASSES.get(sensor.unit),
+    )
+    # percentage unit on battery sensor
+    if sensor.unit == "%" and sensor.kind == SensorKind.BAT:
+        desc.state_class = STATE_CLASS_MEASUREMENT
+        desc.device_class = DEVICE_CLASS_BATTERY
+    if sensor.id_ not in _MAIN_SENSORS:
+        desc.entity_category = "diagnostic"
+    return desc
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the GoodWe inverter from a config entry."""
+    entities = []
+    inverter = hass.data[DOMAIN][config_entry.entry_id][KEY_INVERTER]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+
+    device_info: DeviceInfo = {
+        "configuration_url": "https://www.semsportal.com",
```

Comment:
Please create DeviceInfo directly (not cast from dict)
```suggestion
    device_info = DeviceInfo(
        configuration_url = "https://www.semsportal.com",
```

Reply:
Done